posts/2021-02-11-Rust-1.50.0.md

```diff
@@ -0,0 +1,140 @@
+
+[install]: https://www.rust-lang.org/install.html
+[notes]: https://github.com/rust-lang/rust/blob/master/RELEASES.md#version-1500-2021-02-11
+
+## What's in 1.50.0 stable
+
+For this release, we have improved array indexing, expanded union field safety, refined file descriptors, and added to the standard library.
+See the [detailed release notes][notes] to learn about other changes
+not covered by this post.
+
+### Const-generic array indexing
+
+Continuing the march toward stable `const` generics, this release adds
+implementations of `ops::Index` and `IndexMut` for arrays `[T; N]` for
+_any_ length of `const N`. The indexing operator `[]` already worked on
+arrays through built-in compiler magic, but at the type level, arrays
+didn't actually implement the library traits until now.
+
+```rust
+fn second<C>(container: &C) -> &C::Output
+where
+    C: std::ops::Index<usize> + ?Sized,
+{
+    &container[1]
+}
+
+fn main() {
+    let array: [i32; 3] = [1, 2, 3];
+    assert_eq!(second(&array[..]), &2); // slices worked before
+    assert_eq!(second(&array), &2); // now it also works directly
+}
+```
+
+### Safe assignments to `ManuallyDrop<T>` union fields
+
+Rust 1.49 made it possible to add `ManuallyDrop<T>` fields to a `union`
+as part of allowing `Drop` for unions at all; however, unions don't know
+which variant is valid (they don't drop any old value on assignment), so
+safe Rust has limited this to `Copy` types only, which never `Drop`. Of
+course, that's also true of `ManuallyDrop`, so now Rust 1.50 allows safe
+assignments to these fields as well.
+
+### A niche for `File` on Unix platforms
+
+Some types in Rust have specific limitations on what is considered a
+valid value, which may not cover the entire range of possible memory
+values. We call any remaining invalid value a [niche], and this space
+may be used for type layout optimizations. For example, in Rust 1.28
+we introduced `NonZero` types where `0` is a niche, and this allowed
+`Option<NonZero>` to use `0` to represent `None` with no extra memory.
+
+On Unix platforms, Rust's `File` is simply made of the system's integer
+file descriptor, and this happens to have a possible niche
+as well because it can never be `-1`! System calls which return a file
+descriptor use `-1` to indicate that an error occurred (check `errno`)
+so it's never possible for `-1` to be a real file descriptor. Starting
+in Rust 1.50 this niche is added to the type's definition so it can be
+used in layout optimizations too. It follows that `Option<File>` will
+now have the same size as `File` itself!
+
+[niche]: https://rust-lang.github.io/unsafe-code-guidelines/glossary.html#niche
+
+### Library changes
+
+In Rust 1.50.0, there are nine new stable functions:
+
+- [`bool::then`]
+- [`btree_map::Entry::or_insert_with_key`]
+- [`f32::clamp`]
+- [`f64::clamp`]
+- [`hash_map::Entry::or_insert_with_key`]
+- [`Ord::clamp`]
+- [`RefCell::take`]
+- [`slice::fill`]
```

Comment:
Hm, I am now wondering if this would be a better thing to call out than the File niche - it seems more likely to be notable to most users and is pretty nice as "safe Rust API" where the C ones is pretty annoying to use in a safe way.

Reply:
My muse fails me on this one. I agree `slice::fill` is a useful method, but even the documentation has only a terse sentence and short example. If you're feeling inspired, would you (or anyone else) care to try writing something for this?

Reply:
> is pretty nice as "safe Rust API" where the C ones is pretty annoying to use in a safe way.

This is the same as `memset`, right? Why is this annoying in C? Memset takes the length of the buffer.

Reply:
memset is also used to e.g. zero out arrays of structs, where it's finicky (IME) to get the length and such right in a guaranteed way. this API avoids those problems. but I don't feel strongly.

Reply:
Ah right, because people tend to put `sizeof(T*)` instead of `sizeof(T)`. Ok, makes sense.

Reply:
> Hm, I am now wondering if this would be a better thing to call out than the File niche

FWIW, I don't think it needs to be an either or situation, the blog post as it stands is about 600 words, adding one more paragraph isn't going to significantly increase the length. It takes roughly 3 minutes to read. I don't think I've ever seen someone complain about the blog post being too long, but I have seen people lament it being too short.